File: server/test_db_connection.ts
#!/usr/bin/env tsx
/**
 * Test script for Node.js database connection (db.ts)
 * Tests connection to Cloud SQL or Neon database
 */

import { pool, db } from './db';
import { sql } from 'drizzle-orm';
import fs from 'fs';
import path from 'path';

async function testConnection() {
  console.log('='.repeat(60));
  console.log('🔍 Testing Node.js Database Connection (db.ts)');
  console.log('='.repeat(60));
  console.log();

  // Display configuration
  console.log('📋 Configuration:');
  
  // Show environment
  const nodeEnv = (process.env.NODE_ENV || '').toLowerCase();
  console.log(`   NODE_ENV: ${nodeEnv || '(not set)'}`);
  
  // Show which database URL variables are set
  const databaseUrlDev = process.env.DATABASE_URL_DEV;
  const databaseUrlProd = process.env.DATABASE_URL_PROD;
  const databaseUrlFallback = process.env.DATABASE_URL;
  
  console.log();
  console.log('   Environment Variables:');
  if (databaseUrlDev) {
    const displayDev = databaseUrlDev.length > 50 ? databaseUrlDev.substring(0, 47) + '...' : databaseUrlDev;
    console.log(`   ✅ DATABASE_URL_DEV: ${displayDev}`);
  } else {
    console.log(`   ⚪ DATABASE_URL_DEV: (not set)`);
  }
  
  if (databaseUrlProd) {
    const displayProd = databaseUrlProd.length > 50 ? databaseUrlProd.substring(0, 47) + '...' : databaseUrlProd;
    console.log(`   ✅ DATABASE_URL_PROD: ${displayProd}`);
  } else {
    console.log(`   ⚪ DATABASE_URL_PROD: (not set)`);
  }
  
  if (databaseUrlFallback) {
    const displayFallback = databaseUrlFallback.length > 50 ? databaseUrlFallback.substring(0, 47) + '...' : databaseUrlFallback;
    console.log(`   ✅ DATABASE_URL: ${displayFallback}`);
  } else {
    console.log(`   ⚪ DATABASE_URL: (not set)`);
  }
  
  // Determine which one is being used (same logic as db.ts)
  let databaseUrl: string;
  if (nodeEnv === 'development') {
    databaseUrl = databaseUrlDev || databaseUrlFallback || '';
  } else if (nodeEnv === 'production') {
    databaseUrl = databaseUrlProd || databaseUrlFallback || '';
  } else {
    databaseUrl = databaseUrlFallback || '';
  }
  
  console.log();
  console.log('   Selected Database URL:');
  if (nodeEnv === 'development') {
    if (databaseUrlDev) {
      console.log(`   🔵 Using: DATABASE_URL_DEV (NODE_ENV=development)`);
    } else if (databaseUrlFallback) {
      console.log(`   🔵 Using: DATABASE_URL (fallback, NODE_ENV=development)`);
    } else {
      console.log(`   ❌ No database URL available for development`);
    }
  } else if (nodeEnv === 'production') {
    if (databaseUrlProd) {
      console.log(`   🔵 Using: DATABASE_URL_PROD (NODE_ENV=production)`);
    } else if (databaseUrlFallback) {
      console.log(`   🔵 Using: DATABASE_URL (fallback, NODE_ENV=production)`);
    } else {
      console.log(`   ❌ No database URL available for production`);
    }
  } else {
    if (databaseUrlFallback) {
      console.log(`   🔵 Using: DATABASE_URL (NODE_ENV not set)`);
    } else {
      console.log(`   ❌ No database URL available`);
    }
  }
  
  // Show the actual URL being used (masked)
  if (databaseUrl) {
    let displayUrl = databaseUrl.length > 50 ? databaseUrl.substring(0, 47) + '...' : databaseUrl;
    // Mask password in URL
    if (displayUrl.includes('@') && displayUrl.includes('://')) {
      const parts = displayUrl.split('@');
      if (parts.length === 2) {
        const protocolUserPass = parts[0];
        if (protocolUserPass.includes(':')) {
          const protocolUser = protocolUserPass.split(':')[0] + ':***';
          displayUrl = protocolUser + '@' + parts[1];
        }
      }
    }
    console.log(`   URL: ${displayUrl}`);
  }

  // Check SSL configuration
  const dbUrlLower = databaseUrl.toLowerCase();
  const isNeon = dbUrlLower.includes('neon.tech');
  const isCloudSQL = 
    process.env.DB_SSL_DIR || 
    process.env.DB_SSL_ROOT_CERT || 
    process.env.DB_SSL_CERT || 
    process.env.DB_SSL_KEY ||
    dbUrlLower.includes('cloudsql') ||
    dbUrlLower.includes('gcp');

  console.log();
  if (isNeon) {
    const envLabel = nodeEnv === 'production' ? 'PRODUCTION' : 'DEVELOPMENT';
    console.log(`   Database Type: 🔵 Neon (${envLabel})`);
  } else if (isCloudSQL) {
    console.log(`   Database Type: 🟢 Cloud SQL`);
  } else {
    console.log(`   Database Type: ⚪ Other`);
  }

  if (isCloudSQL) {
    if (process.env.DB_SSL_DIR) {
      console.log(`   DB_SSL_DIR: ${process.env.DB_SSL_DIR}`);
      const sslDir = process.env.DB_SSL_DIR;
      const certs = ['server-ca.pem', 'client-cert.pem', 'client-key.pem'];
      certs.forEach(cert => {
        const certPath = path.join(sslDir, cert);
        if (fs.existsSync(certPath)) {
          console.log(`   ✅ Found: ${cert}`);
        } else {
          console.log(`   ⚠️  Missing: ${cert}`);
        }
      });
    } else {
      if (process.env.DB_SSL_ROOT_CERT) {
        const exists = fs.existsSync(process.env.DB_SSL_ROOT_CERT);
        console.log(`   DB_SSL_ROOT_CERT: ${process.env.DB_SSL_ROOT_CERT}`);
        console.log(`   ${exists ? '✅' : '❌'} File exists`);
      }
      if (process.env.DB_SSL_CERT) {
        const exists = fs.existsSync(process.env.DB_SSL_CERT);
        console.log(`   DB_SSL_CERT: ${process.env.DB_SSL_CERT}`);
        console.log(`   ${exists ? '✅' : '❌'} File exists`);
      }
      if (process.env.DB_SSL_KEY) {
        const exists = fs.existsSync(process.env.DB_SSL_KEY);
        console.log(`   DB_SSL_KEY: ${process.env.DB_SSL_KEY}`);
        console.log(`   ${exists ? '✅' : '❌'} File exists`);
      }
    }
  }

  console.log();
  console.log('-'.repeat(60));
  console.log();

  // Test connection
  console.log('🔌 Testing Connection...');
  try {
    // Test 1: Basic connection test
    const client = await pool.connect();
    try {
      const result = await client.query('SELECT 1 as test');
      console.log(`   ✅ Basic query test passed (result: ${result.rows[0].test})`);

      // Test 2: Get PostgreSQL version
      const versionResult = await client.query('SELECT version()');
      const version = versionResult.rows[0].version.split(',')[0];
      console.log(`   ✅ PostgreSQL version: ${version}`);

      // Test 3: Get current database name
      const dbNameResult = await client.query('SELECT current_database()');
      console.log(`   ✅ Connected to database: ${dbNameResult.rows[0].current_database}`);

      // Test 4: Get current user
      const userResult = await client.query('SELECT current_user');
      console.log(`   ✅ Connected as user: ${userResult.rows[0].current_user}`);

      // Test 5: Test SSL connection
      console.log();
      console.log('🔒 Testing SSL Connection...');
      try {
        const sslResult = await client.query('SHOW ssl');
        console.log(`   SSL Status: ${sslResult.rows[0].ssl}`);
      } catch (e) {
        // Some databases don't support SHOW ssl
        console.log('   ℹ️  SSL status query not available');
      }

      // Test 6: Test Drizzle ORM connection
      console.log();
      console.log('📊 Testing Drizzle ORM Connection...');
      const drizzleResult = await db.execute(sql`SELECT 1 as test`);
      console.log(`   ✅ Drizzle ORM query successful`);

      // Test 7: List tables
      console.log();
      console.log('📋 Testing Database Schema...');
      const tablesResult = await client.query(`
        SELECT table_name 
        FROM information_schema.tables 
        WHERE table_schema = 'public' 
        ORDER BY table_name
        LIMIT 20
      `);

      if (tablesResult.rows.length > 0) {
        console.log(`   ✅ Found ${tablesResult.rows.length} tables:`);
        tablesResult.rows.forEach((row, idx) => {
          if (idx < 10) {
            console.log(`      - ${row.table_name}`);
          }
        });
        if (tablesResult.rows.length > 10) {
          console.log(`      ... and ${tablesResult.rows.length - 10} more`);
        }
      } else {
        console.log('   ⚠️  No tables found in public schema');
      }

      // Test 8: Connection pool stats
      console.log();
      console.log('📈 Connection Pool Stats:');
      console.log(`   Total connections: ${pool.totalCount}`);
      console.log(`   Idle connections: ${pool.idleCount}`);
      console.log(`   Waiting clients: ${pool.waitingCount}`);

    } finally {
      client.release();
    }

    console.log();
    console.log('='.repeat(60));
    console.log('✅ All connection tests passed!');
    console.log('='.repeat(60));
    return true;

  } catch (error: any) {
    console.log();
    console.log('='.repeat(60));
    console.log('❌ Connection test failed!');
    console.log('='.repeat(60));
    console.log(`Error: ${error.message}`);
    if (error.stack) {
      console.log();
      console.log('Stack trace:');
      console.log(error.stack);
    }
    console.log();
    console.log('Troubleshooting tips:');
    console.log('1. Verify DATABASE_URL, DATABASE_URL_DEV, or DATABASE_URL_PROD is correct');
    console.log('2. Check that NODE_ENV is set to \'development\' or \'production\'');
    console.log('3. Check that SSL certificates are in the correct location (for Cloud SQL)');
    console.log('4. Ensure the database instance allows connections from your IP');
    console.log('5. Verify database credentials are correct');
    console.log('6. Check firewall rules');
    return false;
  } finally {
    // Close the pool
    await pool.end();
  }
}

// Run the test
testConnection()
  .then((success) => {
    process.exit(success ? 0 : 1);
  })
  .catch((error) => {
    console.error('Fatal error:', error);
    process.exit(1);
  });

File: python_backend/test_db_connection.py
#!/usr/bin/env python3
"""
Test script for database connection
Tests the database connection with environment-based URL selection.
"""

import asyncio
import os
import sys
from pathlib import Path

# Add the src directory to the path
sys.path.insert(0, str(Path(__file__).parent / "src"))

from database.connection import get_db_pool, close_db_pool, DATABASE_URL, DB_SSL_DIR, DB_SSL_ROOT_CERT, DB_SSL_CERT, DB_SSL_KEY


async def test_connection():
    """Test database connection and perform basic queries"""
    print("=" * 60)
    print("🔍 Testing Database Connection")
    print("=" * 60)
    print()
    
    # Display configuration
    print("📋 Configuration:")
    
    # Show environment
    node_env = os.getenv("NODE_ENV", "").lower()
    print(f"   NODE_ENV: {node_env or '(not set)'}")
    
    # Show which database URL variables are set
    database_url_dev = os.getenv("DATABASE_URL_DEV")
    database_url_prod = os.getenv("DATABASE_URL_PROD")
    database_url_fallback = os.getenv("DATABASE_URL")
    
    print()
    print("   Environment Variables:")
    if database_url_dev:
        display_dev = database_url_dev[:50] + "..." if len(database_url_dev) > 50 else database_url_dev
        print(f"   ✅ DATABASE_URL_DEV: {display_dev}")
    else:
        print(f"   ⚪ DATABASE_URL_DEV: (not set)")
    
    if database_url_prod:
        display_prod = database_url_prod[:50] + "..." if len(database_url_prod) > 50 else database_url_prod
        print(f"   ✅ DATABASE_URL_PROD: {display_prod}")
    else:
        print(f"   ⚪ DATABASE_URL_PROD: (not set)")
    
    if database_url_fallback:
        display_fallback = database_url_fallback[:50] + "..." if len(database_url_fallback) > 50 else database_url_fallback
        print(f"   ✅ DATABASE_URL: {display_fallback}")
    else:
        print(f"   ⚪ DATABASE_URL: (not set)")
    
    # Determine which one is being used
    print()
    print("   Selected Database URL:")
    if node_env == "development":
        if database_url_dev:
            print(f"   🔵 Using: DATABASE_URL_DEV (NODE_ENV=development)")
        elif database_url_fallback:
            print(f"   🔵 Using: DATABASE_URL (fallback, NODE_ENV=development)")
        else:
            print(f"   ❌ No database URL available for development")
    elif node_env == "production":
        if database_url_prod:
            print(f"   🔵 Using: DATABASE_URL_PROD (NODE_ENV=production)")
        elif database_url_fallback:
            print(f"   🔵 Using: DATABASE_URL (fallback, NODE_ENV=production)")
        else:
            print(f"   ❌ No database URL available for production")
    else:
        if database_url_fallback:
            print(f"   🔵 Using: DATABASE_URL (NODE_ENV not set)")
        else:
            print(f"   ❌ No database URL available")
    
    # Show the actual URL being used (masked)
    if DATABASE_URL:
        display_url = DATABASE_URL[:50] + "..." if len(DATABASE_URL) > 50 else DATABASE_URL
        # Mask password in URL
        if "@" in display_url and "://" in display_url:
            parts = display_url.split("@")
            if len(parts) == 2:
                protocol_user_pass = parts[0]
                if ":" in protocol_user_pass:
                    protocol_user = protocol_user_pass.split(":")[0] + ":***"
                    display_url = protocol_user + "@" + parts[1]
        print(f"   URL: {display_url}")
    
    # Detect database type
    db_url_lower = DATABASE_URL.lower() if DATABASE_URL else ""
    is_neon = 'neon.tech' in db_url_lower
    is_cloud_sql = (
        'cloudsql' in db_url_lower or 
        'gcp' in db_url_lower or
        DB_SSL_ROOT_CERT or 
        DB_SSL_CERT or 
        DB_SSL_KEY or
        DB_SSL_DIR
    )
    
    print()
    if is_neon:
        env_label = "PRODUCTION" if node_env == "production" else "DEVELOPMENT"
        print(f"   Database Type: 🔵 Neon ({env_label})")
    elif is_cloud_sql:
        print(f"   Database Type: 🟢 Cloud SQL")
    else:
        print(f"   Database Type: ⚪ Other")
    
    if DB_SSL_DIR:
        print(f"   DB_SSL_DIR: {DB_SSL_DIR}")
        ssl_dir = Path(DB_SSL_DIR)
        if ssl_dir.exists():
            print(f"   ✅ SSL directory exists")
            for cert_file in ["server-ca.pem", "client-cert.pem", "client-key.pem"]:
                cert_path = ssl_dir / cert_file
                if cert_path.exists():
                    print(f"   ✅ Found: {cert_file}")
                else:
                    print(f"   ⚠️  Missing: {cert_file}")
        else:
            print(f"   ❌ SSL directory does not exist")
    elif DB_SSL_ROOT_CERT or DB_SSL_CERT or DB_SSL_KEY:
        if DB_SSL_ROOT_CERT:
            print(f"   DB_SSL_ROOT_CERT: {DB_SSL_ROOT_CERT}")
            print(f"   {'✅' if Path(DB_SSL_ROOT_CERT).exists() else '❌'} File exists")
        if DB_SSL_CERT:
            print(f"   DB_SSL_CERT: {DB_SSL_CERT}")
            print(f"   {'✅' if Path(DB_SSL_CERT).exists() else '❌'} File exists")
        if DB_SSL_KEY:
            print(f"   DB_SSL_KEY: {DB_SSL_KEY}")
            print(f"   {'✅' if Path(DB_SSL_KEY).exists() else '❌'} File exists")
    else:
        if not is_neon:
            print("   ⚠️  No SSL certificates configured")
    
    print()
    print("-" * 60)
    print()
    
    # Test connection
    print("🔌 Testing Connection...")
    try:
        pool = await get_db_pool()
        print("   ✅ Connection pool created successfully")
        
        # Test basic query
        async with pool.acquire() as conn:
            result = await conn.fetchval("SELECT 1")
            print(f"   ✅ Basic query test passed (result: {result})")
            
            # Get PostgreSQL version
            version = await conn.fetchval("SELECT version()")
            print(f"   ✅ PostgreSQL version: {version.split(',')[0]}")
            
            # Get current database name
            db_name = await conn.fetchval("SELECT current_database()")
            print(f"   ✅ Connected to database: {db_name}")
            
            # Get current user
            user = await conn.fetchval("SELECT current_user")
            print(f"   ✅ Connected as user: {user}")
            
            # Test listing tables
            print()
            print("📊 Testing Database Schema...")
            tables = await conn.fetch("""
                SELECT table_name 
                FROM information_schema.tables 
                WHERE table_schema = 'public' 
                ORDER BY table_name
            """)
            
            if tables:
                print(f"   ✅ Found {len(tables)} tables:")
                for table in tables[:10]:  # Show first 10 tables
                    print(f"      - {table['table_name']}")
                if len(tables) > 10:
                    print(f"      ... and {len(tables) - 10} more")
            else:
                print("   ⚠️  No tables found in public schema")
            
            # Test SSL connection info
            print()
            print("🔒 Testing SSL Connection...")
            try:
                ssl_info = await conn.fetchval("SHOW ssl")
                print(f"   SSL Status: {ssl_info}")
            except:
                print("   ℹ️  SSL status query not available")
            
            # Try to get SSL version if available
            try:
                ssl_version = await conn.fetchval("SHOW ssl_version")
                print(f"   SSL Version: {ssl_version}")
            except:
                pass
        
        print()
        print("=" * 60)
        print("✅ All connection tests passed!")
        print("=" * 60)
        return True
        
    except Exception as e:
        print()
        print("=" * 60)
        print("❌ Connection test failed!")
        print("=" * 60)
        print(f"Error: {e}")
        print()
        print("Troubleshooting tips:")
        print("1. Verify DATABASE_URL, DATABASE_URL_DEV, or DATABASE_URL_PROD is correct")
        print("2. Check that NODE_ENV is set to 'development' or 'production'")
        print("3. Check that SSL certificates are in the correct location (for Cloud SQL)")
        print("4. Ensure the database instance allows connections from your IP")
        print("5. Verify database credentials are correct")
        print("6. Check firewall rules")
        return False
    
    finally:
        await close_db_pool()
        print("   ✅ Connection pool closed")


async def main():
    """Main test function"""
    if not DATABASE_URL:
        print("❌ Error: No database URL available")
        print()
        print("Please set one of the following:")
        print("  export DATABASE_URL='postgresql://user:pass@host:port/dbname'")
        print("  export DATABASE_URL_DEV='postgresql://user:pass@host:port/dbname' (for development)")
        print("  export DATABASE_URL_PROD='postgresql://user:pass@host:port/dbname' (for production)")
        print("  export NODE_ENV='development' or 'production'")
        sys.exit(1)
    
    success = await test_connection()
    sys.exit(0 if success else 1)


if __name__ == "__main__":
    asyncio.run(main())
